### PR TITLE
Use Bazel to build the entire backend and perform API code generation

### PR DESCRIPTION
**Pull request by neuromage**
Using Bazel means we'll get reproducible builds that are not dependent on the tools installed by each dev. This change adds BUILD files for the backend that builds both binaries, as well as the API protocol buffers and swagger definitions.

In order to let `go build` and `go test` work as before, the Makefile is replaced with a script that uses Bazel to first generate the API code, and then copy them back into the source tree.

Most of the BUILD files were generated automatically using Gazelle.

Apologies for the large PR but this is hopefully a one time change. From here on, running `bazel //:gazelle` should update the BUILD files automatically. 

We may also want to consider not checking in generated code altogether. This may be a better option going forward.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/609)

### CODE REVIEW

**Comment by neuromage** [comment]
Note: the changes in the generated files are due to using more up to date versions of the code generation tools.

File: backend/api/experiment.proto [hunk 1]
```diff
@@ -20,6 +20,7 @@ import "google/api/annotations.proto";
 import "google/protobuf/empty.proto";
 import "google/protobuf/timestamp.proto";
 import "protoc-gen-swagger/options/annotations.proto";
+import "backend/api/error.proto";
```
Comment (IronPan) [inline backend/api/experiment.proto]: is this change required or by accident?

File: backend/api/go_client/BUILD.bazel [hunk 2]
```diff
@@ -0,0 +1,39 @@
+load("@io_bazel_rules_go//go:def.bzl", "go_library")
+
+go_library(
+    name = "go_default_library",
+    srcs = [
+        "error.pb.go",
```
Comment (IronPan) [inline backend/api/go_client/BUILD.bazel]: can this be wildcard *.pb.go *.pb.gw.go? otherwise this need to manually added in the future i guess?

File: backend/api/go_client/error.pb.go [hunk 3]
```diff
@@ -1,28 +1,12 @@
-// Copyright 2018 Google LLC
```
Comment (IronPan) [inline backend/api/go_client/error.pb.go]: headers seems gone here

**Comment by neuromage** [comment]
I added a README with instructions on how to build/test with bazel. The BUILD files don't need to be manually maintained, for the most part, running `bazel run //:gazelle` will auto-update them.

As for whether we should keep checking generated code in, here are the pros and cons as I see them:

Pros:
* `go build` and `go test` continue to work as before. No need to change build/release process as yet.
* IDE/Editors see the generated code, and don't need special Bazel plugins to auto-complete on them.

Cons;
* Generated code changes are noisy during code reviews
* Spurious merge conflicts to resolve when changes touch proto files.

If we do decide to remove them, my preference is to do that in a separate PR.

**Comment by yebrahim** [comment]
Thanks Ajay. Looks like there's still some generated files missing the license headers.
Is there value in checking in the BUILD files if they're all auto-generated? Can the gazelle build rule be run before running the rest of the bazel rules? This way we can just add them to .gitignore.

**Comment by yebrahim** [comment]
I guess it's a good idea to discuss maintaining auto-generated files in a separate issue so as not to block this PR on it, but it's good to make sure we really need any new generated files.

**Comment by yebrahim** [comment]
SG then.
/lgtm

File: backend/api/experiment.proto (hunk 1 above)
Comment (IronPan) [reply backend/api/experiment.proto]: thanks for confirm.
